pkg/ebpf/generic_map.go

```diff
@@ -0,0 +1,352 @@
+	}
+	gm, err := Map[K, V](m)
+	if err != nil {
+		return nil, err
+	}
+	return gm, nil
+}
+
+// Map returns the underlying ebpf.Map
+func (g *GenericMap[K, V]) Map() *ebpf.Map {
+	return g.m
+}
+
+// IteratorOptions are options for the Iterate method
+type IteratorOptions struct {
+	BatchSize       int  // Number of items to fetch per batch. If 0, use default value (100)
+	ForceSingleItem bool // Force the use of the single item iterator even if the batch API is supported
+}
+
+// Put inserts a new key/value pair in the map. If the key already exists, the value is updated
+func (g *GenericMap[K, V]) Put(key *K, value *V) error {
+	if g.isPerCPU() {
+		return g.m.Put(unsafe.Pointer(key), *value)
+	}
+
+	return g.m.Put(unsafe.Pointer(key), unsafe.Pointer(value))
+}
+
+// Update updates the value of an existing key in the map.
+func (g *GenericMap[K, V]) Update(key *K, value *V, flags ebpf.MapUpdateFlags) error {
+	return g.m.Update(unsafe.Pointer(key), unsafe.Pointer(value), flags)
+}
+
+// Lookup looks up a key in the map and returns the value. If the key doesn't exist, it returns ErrKeyNotExist
+func (g *GenericMap[K, V]) Lookup(key *K, valueOut *V) error {
+	if g.isPerCPU() {
+		return g.m.Lookup(unsafe.Pointer(key), *valueOut)
+	}
+
+	return g.m.Lookup(unsafe.Pointer(key), unsafe.Pointer(valueOut))
+}
+
+// Delete deletes a key from the map. If the key doesn't exist, it returns ErrKeyNotExist
+func (g *GenericMap[K, V]) Delete(key *K) error {
+	return g.m.Delete(unsafe.Pointer(key))
+}
+
+// BatchDelete deletes a batch of keys from the map. Returns the number of deleted items
+func (g *GenericMap[K, V]) BatchDelete(keys []K) (int, error) {
+	return g.m.BatchDelete(keys, nil)
+}
+
+// BatchUpdate updates a batch of keys in the map
+func (g *GenericMap[K, V]) BatchUpdate(keys []K, values []V, opts *ebpf.BatchOptions) (int, error) {
+	return g.m.BatchUpdate(keys, values, opts)
+}
+
+// GenericMapIterator is an interface for iterating over a GenericMap
+type GenericMapIterator[K interface{}, V interface{}] interface {
+	// Next fills K and V with the next key/value pair in the map. It returns false if there are no more elements
+	Next(key *K, value *V) bool
+
+	// Err returns the last error that happened during iteration.
+	Err() error
+}
+
+func isPerCPU(t ebpf.MapType) bool {
+	switch t {
+	case ebpf.PerCPUHash, ebpf.PerCPUArray, ebpf.LRUCPUHash:
+		return true
+	}
+	return false
+}
+
+func (g *GenericMap[K, V]) isPerCPU() bool {
+	return isPerCPU(g.m.Type())
+}
+
+// Iterate returns an iterator for the map, which transparently chooses between batch and single item
+func (g *GenericMap[K, V]) Iterate() GenericMapIterator[K, V] {
+	return g.IterateWithOptions(IteratorOptions{BatchSize: defaultBatchSize})
+}
+
+func (g *GenericMap[K, V]) valueTypeCanUseUnsafePointer() bool {
+	// Simple test for now, but we probably will need to add more cases,
+	// as I am not 100% sure of the behavior of structs with maps
```

Comment:
structs are OK and can use `unsafe.Pointer`. That is what we primarily use at the moent.